### ISSUE TITLE
Automate code formatting

### ISSUE THREAD

**Issue by tsalo** [ME-ICA/tedana#219]
### Summary
This is just something that came up recently in the [python_pals](https://brainhack.slack.com/messages/C2Y1T5XDJ) BrainHack Slack channel, although I think we've discussed it before informally. If we use [Black](https://black.readthedocs.io/en/stable/), we can reduce the amount of mental overhead we need for dealing with coding style. Not having to worry about style will also make it easier for new contributors.

### Next Steps
1. Add instructions for using Black to contributing guidelines. All contributors should have to do is run it on the package prior to opening a PR.
2. Run Black on the package and merge in a new PR to get things started.

**Comment by emdupre** [comment]
Thanks for opening this ! Black seems like another opinionated code formatter, similar to [Prettier](https://prettier.io/) for javascript. Code formatters can be great things, for many of the reasons you mentioned: reduced mental overhead and (enforced) consistent styling across contributors, among others.

I personally think these are especially valuable when programmers all have about the same level of experience, but different stylistic preferences. So if we both follow PEP8 in slightly different ways (i.e., double vs single quotes) and we want the code base to look unified, then they have a great use.

My concern is actually that, as an open source project, I think one of the things we can offer to new contributors is a chance to learn what PEP8 (and code styling, more generally) is. We enforce linting with Travis, so I'm not worried that major PEP8 violations are sneaking in and making the code base harder to read. And, personally, I'm ok with alternating between double and single quotes in the same file.

My suggestion would be: if people want to run their code through Black, we can provide instructions on how to do so. It shouldn't affect our linting, since that's the whole point of a code formatter :laughing: But I'm very hesitant to take this on as our default.

**Comment by tsalo** [comment]
Black really only seems to change whitespace, how lines are split, and the quotes. I think there's a lot of good coding practice that newcomers can still learn, even if we eliminate the need to manage those aspects. Using Black won't be a substitute for good commenting, using informative variable names, or using good coding techniques. Still, if we do want to make it optional, then we should at least do it once (i.e., #220) to prevent someone else later on from opening a PR that mixes content with style. I can update the contributing instructions there to make sure contributors know that running Black is optional.

I did notice that Black was making one error repeatedly, when merging strings across multiple lines. See the changes I made in [this commit](https://github.com/ME-ICA/tedana/pull/220/commits/b62596a3cd5937feca5f487f73dc021c37b912a7) for examples. It's not a big deal, but it's something we'll have to check when we review PRs.

**Comment by jbteves** [comment]
As a newcomer myself, I'll comment that reviews are much more helpful to my contribution development, and code formatting is a total afterthought to me, so I wouldn't feel like my experience is slighted by using auto-formatting. 
Granted, formatting is always an afterthought to me since I'm bouncing between 3-4 languages at any time.

**Comment by jbteves** [comment]
There hasn't been any recent conversation on this. My opinion has evolved somewhat after opening a few more PRs, and I think that the auto-formatting may actually be less seamless than simply opening a PR and correcting what Travis doesn't like. It's one fewer package to manage, and it seems conceptually like it could be more bug-prone than simply checking rules. It is also a little redundant to Travis. If you have a strong opinion about it @tsalo then we should continue discussion. I would also really like for @dowdlelt to chime in; do you have an opinion? You are the most recent contributor other than myself to the codebase.

**Comment by jbteves** [comment]
This hasn't really gained any traction; closing. @tsalo if you disagree we can reopen it.

**Comment by jbteves** [comment]
It looks like `black` is getting to be a common formatter in use by several of the dev team, including @tsalo and @eurunuela. Consequently, I think this is worth re-opening and revisiting. I am open to switch to using `black` style so that we have consistent formats across developers, however, I would like to discuss a roadmap for applying it.

**Comment by handwerkerd** [comment]
As a bit more background, @jbteves and I were talking about a few recent PRs that were also full of format corrections. That causes headaches when trying to keep other PRs aligned with Main. If we do want to more aggressively mandate formatting, perhaps, after #591 is merged, we can plan to go through the existing code base to apply `black` formatting as distinct PRs so that subsequent PRs are cleaner. This doesn't have to be a huge discussion, but I've added it to the agenda for this Friday's dev call #728

**Comment by eurunuela** [comment]
I agree it makes the reviewing process much harder. I have started using emacs and I have it set up so that it automatically formats the code every time I save it. I have to admit I completely forgot about it as it's not something I consciously do. I am in favor of switching to using `black` after #591 is merged.

**Comment by tsalo** [comment]
Per the minutes from May:

>ACTION ITEM: Plan to rip the style re-formatting band-aid off in one PR after metric modularization is merged

So does that mean we should just run black and isort?

**Comment by jbteves** [comment]
That looks like a good proposal to me.
